# String Formatting

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

    player = 'Sowjanya'
    points = 33
    
    'Last night, '+player+' scored '+str(points)+' points.'  # concatenation
    
    f'Last night, {player} scored {points} points.'          # string formatting


There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

Since you will likely encounter all three versions in someone else's code, we describe each of them here.

## Formatting with placeholders
You can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".

In [1]:
print("I'm going to inject name of %s here." %'satya')

I'm going to inject name of satya here.


In [1]:
#formatting with placeholders
print("My name is %s." %'Barry Allen')

My name is Barry Allen.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [2]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


In [2]:
print("My name is %s and I'm the %s alive." %('Barry Allen','fastest man'))

My name is Barry Allen and I'm the fastest man alive.


You can also pass variable names:

In [3]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


In [5]:
a,b = 'The', 'Flash'
print("My name is %s and I'm the %s alive.I am %s %s." %('Barry Allen','fastest man',a,b))

My name is Barry Allen and I'm the fastest man alive.I am The Flash.


### Format conversion methods.
It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [4]:
print('He said his name was %s.' %'shivendra')
print('She said her name was %r.' %'Nisha')

He said his name was shivendra.
She said her name was 'Nisha'.


In [6]:
print('Madness is like %s.' %'gravity')
print('All it takes is a little %r.' %'push')

Madness is like gravity.
All it takes is a little 'push'.


As another example, `\t` inserts a tab into a string.

In [ ]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this  big')

In [9]:
print('For Uxas!For %s.' %' \t Darkseid')
print('For Uxas!For %r.' %'Darkseid')

For Uxas!For  	 Darkseid.
For Uxas!For 'Darkseid'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first, without rounding. Note the difference below:

In [6]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


In [11]:
print('The Dark Knight has an Imdb rating of %s.' %9.00)
print('The Dark Knight has an Imdb rating of %d.' %9.00)

The Dark Knight has an Imdb rating of 9.0.
The Dark Knight has an Imdb rating of 9.


### Padding and Precision of Floating Point Numbers
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [7]:
print('Floating point numbers: %6.2f' %(13.144))

Floating point numbers:  13.14


In [14]:
print('Floating point numbers: %6.2f' %(12.897))

Floating point numbers:  12.90


In [8]:
print('Floating point numbers: %1.3f' %(13.144))

Floating point numbers: 13.144


In [15]:
print('Floating point numbers: %1.3f' %(12.897))

Floating point numbers: 12.897


In [9]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [16]:
print('Floating point numbers: %1.5f' %(12.897))

Floating point numbers: 12.89700


In [10]:
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [17]:
print('Floating point numbers: %10.2f' %(12.897))

Floating point numbers:      12.90


In [11]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


In [19]:
print('Floating point numbers: %25.2f' %(12.897))

Floating point numbers:                     12.90


For more information on string formatting with placeholders visit https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Multiple Formatting
Nothing prohibits using more than one conversion tool in the same print statement:

In [12]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


In [26]:
print('Superman: %s, The League: %5.2f, Wonder Woman: %r' %('Kryptonian',6.7592,'Amazon'))

Superman: Kryptonian, The League:  6.76, Wonder Woman: 'Amazon'


## Formatting with the `.format()` method
A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [13]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


In [21]:
print('Winter is {}'.format('coming'))

Winter is coming


### The .format() method has several advantages over the %s placeholder method:

#### 1. Inserted objects can be called by index position:

In [24]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


In [25]:
print('{2} {0} {1}'.format('The','Martian','Manhunter'))

Manhunter The Martian


#### 2. Inserted objects can be assigned keywords:

In [28]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


In [29]:
print('Central City: {a}, Gotham City: {b}, Mother Box: {c}'.format(a='Barry Allen',b='Bruce Wayne',c=03.50))

Central City: Barry Allen, Gotham City: Bruce Wayne, Mother Box: 3.5


#### 3. Inserted objects can be reused, avoiding duplication:

In [16]:
print('A %s saved is a %s earned.' %('rupee','rupee'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='rupee'))

A rupee saved is a rupee earned.
A rupee saved is a rupee earned.


In [31]:
print('A %s in need is a %s indeed.' %('friend','friend'))
# vs.
print('A {p} in need is a {p} indeed.'.format(p='friend'))

A friend in need is a friend indeed.
A friend in need is a friend indeed.


### Alignment, padding and precision with `.format()`
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [17]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


In [35]:
print('{0:5} | {1:15}'.format('Fruit', 'Quantity'))
print('{0:5} | {1:15}'.format('Grapes', 2.))
print('{0:5} | {1:15}'.format('Lemons', 5))

Fruit | Quantity       
Grapes |             2.0
Lemons |               5


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [18]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


In [39]:
print('{0:<3} | {1:^7} | {2:>12}'.format('Left','Center','Right'))
print('{0:<3} | {1:^7} | {2:>12}'.format(11,22,33))

Left | Center  |        Right
11  |   22    |           33


You can precede the aligment operator with a padding character

In [19]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


In [40]:
print('{0:*<8} | {1:+^8} | {2:%>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:&>8}'.format(11,22,33))

Left**** | +Center+ | %%%Right
11====== | ---22--- | &&&&&&33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [20]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


In [41]:
print('This is my ten-character, two-decimal number:%12.8f' %13.579)
print('This is my ten-character, two-decimal number:{0:12.8f}'.format(13.579))

This is my ten-character, two-decimal number: 13.57900000
This is my ten-character, two-decimal number: 13.57900000


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

## Formatted String Literals (f-strings)

Introduced in Python 3.6, f-strings offer several benefits over the older `.format()` string method described above. For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

In [21]:
name = 'Satya'

print(f"He said his name is {name}.")

He said his name is Satya.


In [32]:
x = 'Wolverine'

print(f"They call Weapon-X as {x}.")

They call Weapon-X as Wolverine.


Pass `!r` to get the string representation:

In [22]:
print(f"He said his name is {name!r}")

He said his name is 'Satya'


In [33]:
print(f"They call Weapon-X as {x!r}.")

They call Weapon-X as 'Wolverine'.


#### Float formatting follows `"result: {value:{width}.{precision}}"`

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [23]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


In [42]:
num = 23.45678
print("My 10 character, four decimal number is:{0:12.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{12}.{6}}")

My 10 character, four decimal number is:     23.4568
My 10 character, four decimal number is:     23.4568


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [43]:
num = 23.45
print("My 10 character, four decimal number is:{0:9.2f}".format(num))
print(f"My 10 character, four decimal number is:{num:{9}.{3}}")

My 10 character, four decimal number is:    23.45
My 10 character, four decimal number is:     23.4


In [44]:
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [25]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


In [46]:
num = 23.45
print("My 10 character, four decimal number is:{0:11.6f}".format(num))
print(f"My 10 character, four decimal number is:{num:7.54f}")

My 10 character, four decimal number is:  23.450000
My 10 character, four decimal number is:23.449999999999999289457264239899814128875732421875000000


For more info on formatted string literals visit https://docs.python.org/3/reference/lexical_analysis.html#f-strings

That is the basics of string formatting!